source/http_stream.c

```diff
@@ -201,137 +215,84 @@ PyObject *aws_py_http_client_stream_new(PyObject *self, PyObject *args) {
+     * Therefore, we set `HttpClientStream._binding = capsule` now, instead of returning capsule to caller */
+    if (PyObject_SetAttrString(py_stream, "_binding", capsule) == -1) {
+        goto error;
     }
 
-    Py_ssize_t num_headers = PyDict_Size(request_headers);
-    struct aws_array_list headers;
-
-    if (num_headers > 0) {
-        if (aws_array_list_init_dynamic(&headers, allocator, (size_t)num_headers, sizeof(struct aws_http_header))) {
-            PyErr_SetAwsLastError();
-            goto clean_up_headers;
-        }
-
-        PyObject *key, *value;
-        Py_ssize_t pos = 0;
-
-        while (PyDict_Next(request_headers, &pos, &key, &value)) {
-            struct aws_http_header http_header;
-            http_header.name = aws_byte_cursor_from_pystring(key);
-            http_header.value = aws_byte_cursor_from_pystring(value);
-
-            aws_array_list_push_back(&headers, &http_header);
-        }
-
-        request_options.header_array = headers.data;
-        request_options.num_headers = (size_t)num_headers;
-    }
+    struct aws_http_make_request_options request_options = {
+        .self_size = sizeof(request_options),
+        .request = native_request,
+        .on_response_headers = s_on_incoming_headers,
+        .on_response_header_block_done = s_on_incoming_header_block_done,
+        .on_response_body = s_on_incoming_body,
+        .on_complete = s_on_stream_complete,
+        .user_data = stream,
+    };
 
-    PyObject *on_read_body = PyObject_GetAttrString(py_http_request, "_on_read_body");
-    if (on_read_body && on_read_body != Py_None) {
-        stream->on_read_body = on_read_body;
-        Py_XINCREF(on_read_body);
-        request_options.stream_outgoing_body = s_stream_outgoing_body;
+    stream->native = aws_http_connection_make_request(native_connection, &request_options);
+    if (!stream->native) {
+        PyErr_SetAwsLastError();
+        goto error;
     }
 
-    PyObject *on_incoming_body = PyObject_GetAttrString(py_http_request, "_on_incoming_body");
-    if (on_incoming_body && on_incoming_body != Py_None) {
-        stream->on_incoming_body = on_incoming_body;
-        Py_XINCREF(on_incoming_body);
-        request_options.on_response_body = s_on_incoming_response_body;
-    }
+    /* From hereon, nothing will fail */
 
-    stream->received_headers = PyDict_New();
-    request_options.on_response_headers = s_on_incoming_response_headers;
-    request_options.on_response_header_block_done = s_on_incoming_header_block_done;
-    request_options.on_complete = s_on_stream_complete;
-    request_options.user_data = stream;
+    /* Force python self to stay alive until on_complete callback */
+    Py_INCREF(py_stream);
 
-    struct aws_http_stream *http_stream = aws_http_stream_new_client_request(&request_options);
-    aws_array_list_clean_up(&headers);
+    Py_DECREF(capsule);
+    Py_RETURN_NONE;
 
-    if (!http_stream) {
-        goto clean_up_headers;
+error:
+    if (capsule) {
```

Comment:
Don't we need to decref on the proxy here too, if failure happens after we incref?

Reply:
same thing where capsule destructor will do the cleanup